### DISCUSSION TITLE
A form of SetsRequiredMembers taking an array of member names

### DISCUSSION THREAD

**Discussion by NinoFloris** [dotnet/csharplang#6567] [Language Ideas]
Would the community see value in adding a new form of `SetsRequiredMembers` that takes an array of member names to suppress requiredness for?

Sometimes there are (optional) properties that set one or multiple required properties. Ideally having an attribute on the optional property and setting it should be enough to fulfill the required contract.

A good example is where you're dealing with ranges and you want a shorthand, there are probably other good examples.

```cs
record MyRecord
{
    public required object OtherThings { get; init; }

    public required int MinSomething {get; init; }
    public required int MaxSomething {get; init; }

    [SetsRequiredMembers(nameof(MinSomething), nameof(MaxSomething))]
    public int Something 
    {
        init 
        {
            MinSomething = value;
            MaxSomething = value;   
        }
    }
}
```

**Comment by Deeem2031** [comment]
I was searching for this feature and was very surprised to only find a all-or-nothing solution. In my case I tried to create a parent/child structure kind of like this:
```
var parent = new Parent
{
    Name = "meow",
    Children = [new Child
    {
        Parent = parent,
        Stuff = "cat"
    }]
};

public class Parent
{ 
    public required String Name { get; init; }
    public required IEnumerable<Child> Children { get; init; }
}

public class Child
{ 
    public required Parent Parent { get; init; }
    public required String Stuff { get; init; }
}
```

As this obviously doesn't work I added a constructor to `Parent` like this:

```
using System.Diagnostics.CodeAnalysis;

var parent = new Parent(parent => 
    [new Child
    {
        Parent = parent,
        Stuff = "cat"
    }])
{
    Name = "meow"
};

public class Parent
{
    [SetsRequiredMembers]
    public Parent(Func<Parent, IEnumerable<Child>> childrenFactory)
    {
        Children = childrenFactory(this);
    }

    public required String Name { get; init; }
    public required IEnumerable<Child> Children { get; init; }
}

public class Child
{ 
    public required Parent Parent { get; init; }
    public required String Stuff { get; init; }
}
```

Without `SetsRequiredMembers` it doesn't compile, with `SetsRequiredMembers` you still get "Non-nullable property 'Name' must contain a non-null value when exiting constructor.". `SetsRequiredMember(Children)` would be ideal for this.

If you have to explicitly mention each member you initialize in the constructor it also prevents bugs from later adding properties you don't initialize in this constructor. This pitfall is the reason I don't like the existing attribute `SetsRequiredMembers` at all.

The only real solution right now is to scrap `required` and `init` and stuff everything in the constructor:

```
var parent = new Parent(
    "meow",
    parent =>
    [new Child
    {
        Parent = parent,
        Stuff = "cat"
    }]);

public class Parent
{
    public Parent(String name, Func<Parent, IEnumerable<Child>> childrenFactory)
    {
        Name = name;
        Children = childrenFactory(this);
    }

    public String Name { get; }
    public IEnumerable<Child> Children { get; }
}

public class Child
{ 
    public required Parent Parent { get; init; }
    public required String Stuff { get; init; }
}
```
